course_preference.py

```diff
@@ -29,34 +29,32 @@ def save_preference():
     current_year = get_current_year()
 
     if data is None:
-        return make_response("No data received", 500)
```

Comment:
I don't remember the discussions... you call your `save_preference` route using AJAX. Using `flash` will produce an HTTP 200 code as the page will be rendered. You will therefore never show the error to the user.

Either you can get back to error codes or produce json. Then, catch the error message in your ajax call to show it to the user.